File: app/src/main/java/com/example/alarm/BottomSheetFragment.java
package com.example.alarm;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.example.alarm.databinding.FragmentBottomsheetBinding;
import com.example.alarm.model.Alarm;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;


import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class BottomSheetFragment extends BottomSheetDialogFragment {
    private final boolean isAdd;
    private final Alarm oldAlarm;
    private final AlarmViewModel alarmViewModel;
    private final ExecutorService executorService = Executors.newSingleThreadExecutor();
    private final BottomSheetListener bottomSheetListener;
    private FragmentBottomsheetBinding binding;

    public BottomSheetFragment(AlarmViewModel alarmViewModel, boolean isAdd, Alarm oldAlarm, BottomSheetListener bottomSheetListener) {
        this.alarmViewModel = alarmViewModel;
        this.isAdd = isAdd;
        this.oldAlarm = oldAlarm;
        this.bottomSheetListener = bottomSheetListener;
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.timePicker.setIs24HourView(true);
        if (isAdd) {
            binding.timePicker.setHour(7);
            binding.timePicker.setMinute(0);
        } else {
            binding.heading.setText("Изменить будильник");
            binding.timePicker.setHour(oldAlarm.getTimeHours());
            binding.timePicker.setMinute(oldAlarm.getTimeMinutes());
            if (!oldAlarm.getName().equals("default") && !oldAlarm.getName().isEmpty()) {
                binding.signalName.setText(oldAlarm.getName());
            }
        }

        binding.confirmButton.setOnClickListener(v -> {
            if (isAdd) {
                addNewAlarm();
            } else {
                changeAlarm(oldAlarm);
            }
        });

        binding.cancelButton.setOnClickListener(v -> dismiss());
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentBottomsheetBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    private void addNewAlarm() {
        Alarm alarm = new Alarm(0);
        alarm.setTimeHours(binding.timePicker.getHour());
        alarm.setTimeMinutes(binding.timePicker.getMinute());
        alarm.setName(binding.signalName.getText().toString().isEmpty() ? "default" : binding.signalName.getText().toString());
        alarm.setEnabled(true);
        executorService.execute(() -> {
            alarmViewModel.addAlarm(alarm, result -> {
                new Handler(Looper.getMainLooper()).post(() -> {
                    Toast.makeText(requireContext(), result.second, Toast.LENGTH_SHORT).show();
                    if (result.first) {
                        bottomSheetListener.onAddAlarm(alarm);
                    }
                    dismiss();
                });
            });
        });
    }

    private void changeAlarm(Alarm oldAlarm) {
        Alarm alarmNew = new Alarm(oldAlarm.getId());
        alarmNew.setTimeHours(binding.timePicker.getHour());
        alarmNew.setTimeMinutes(binding.timePicker.getMinute());
        alarmNew.setName(binding.signalName.getText().toString().isEmpty() ? "default" : binding.signalName.getText().toString());
        alarmNew.setEnabled(oldAlarm.getEnabled());

        executorService.execute(() -> {
            alarmViewModel.updateAlarm(alarmNew, result -> {
                new Handler(Looper.getMainLooper()).post(() -> {
                    Toast.makeText(requireContext(), result.second, Toast.LENGTH_SHORT).show();
                if(result.first) {
                    bottomSheetListener.onChangeAlarm(oldAlarm, alarmNew);
                }
                dismiss();
                });
            });
        });
    }
}

File: app/src/main/java/com/example/alarm/BottomSheetFragment.kt
package com.example.alarm

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.alarm.databinding.FragmentBottomsheetBinding
import com.example.alarm.model.Alarm
import com.example.alarm.model.AlarmService
import com.example.alarm.model.MyAlarmManager
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

interface BottomSheetListener {
    fun onAddAlarm(alarm: Alarm)
    fun onChangeAlarm(alarmOld: Alarm, alarmNew: Alarm)
}

class BottomSheetFragment(
    private val isAdd: Boolean,
    private val oldAlarm: Alarm,
    private val bottomSheetListener: BottomSheetListener
) : BottomSheetDialogFragment() {

    private lateinit var binding: FragmentBottomsheetBinding

    private val job = Job()
    private val uiScope = CoroutineScope(Dispatchers.Main + job)
    private val alarmsService: AlarmService
        get() = Repositories.alarmRepository as AlarmService

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.timePicker.setIs24HourView(true)
        if(isAdd) {
            binding.timePicker.hour = 7
            binding.timePicker.minute = 0
        }
        else {
            binding.heading.text = "Изменить будильник"
            binding.timePicker.hour = oldAlarm.timeHours
            binding.timePicker.minute = oldAlarm.timeMinutes
            if((oldAlarm.name != "default") && (oldAlarm.name != "")) binding.signalName.setText(oldAlarm.name)
        }
        binding.confirmButton.setOnClickListener {
            if(isAdd) { addNewAlarm() }
            else { changeAlarm(oldAlarm) }
        }
        binding.cancelButton.setOnClickListener {
            dismiss()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentBottomsheetBinding.inflate(inflater, container, false)
        return binding.root
    }

    private fun addNewAlarm() {
        val alarm = Alarm(
            id = 0,
            timeHours = binding.timePicker.hour,
            timeMinutes = binding.timePicker.minute,
            name = if(binding.signalName.text.toString() == "") "default" else binding.signalName.text.toString() ,
            enabled = 1
        )
        uiScope.launch {
            alarmsService.addAlarm(alarm)
            val settings = async { alarmsService.getSettings() }
            MyAlarmManager(context, alarm, settings.await()).startProcess()
            Log.d("testSettingsAddAlarm", settings.await().toString())
            bottomSheetListener.onAddAlarm(alarm)
            dismiss()
        }
    }
    private fun changeAlarm(oldAlarm: Alarm) {
        val alarmNew = Alarm(
            id = oldAlarm.id,
            timeHours = binding.timePicker.hour,
            timeMinutes = binding.timePicker.minute,
            name = if(binding.signalName.text.toString() == "") "default" else binding.signalName.text.toString(),
            enabled = oldAlarm.enabled
        )
        uiScope.launch {
            alarmsService.updateAlarm(alarmNew)
            if (oldAlarm.enabled == 1) {
                val settings = async { alarmsService.getSettings() }
                MyAlarmManager(context, alarmNew, settings.await()).restartProcess()
            }
            bottomSheetListener.onChangeAlarm(oldAlarm, alarmNew)
            dismiss()
        }
    }
}